### PR TITLE
store: fix nil panic in proxy heap

### PR DESCRIPTION
**Pull request by GiedriusS**
With lazy proxying we need to wait for at least one response before we can build a heap properly.

Closes https://github.com/thanos-io/thanos/pull/5717 Fixes https://github.com/thanos-io/thanos/pull/5552.

Improved e2e tests to cover this case. cc @matej-g @fpetkovski 

Signed-off-by: Giedrius Statkevičius <giedrius.statkevicius@vinted.com>

### CODE REVIEW

File: pkg/store/proxy_heap.go [hunk 1]
```diff
@@ -293,6 +293,19 @@ func (l *lazyRespSet) Empty() bool {
 	l.bufferedResponsesMtx.Lock()
 	defer l.bufferedResponsesMtx.Unlock()
 
+	// NOTE(GiedriusS): need to wait here for at least one
+	// response so that we could build the heap properly.
+	if l.noMoreData && len(l.bufferedResponses) == 0 {
+		return true
+	}
+
+	for len(l.bufferedResponses) == 0 {
```
Comment (fpetkovski) [inline pkg/store/proxy_heap.go:302]: Do we need to unlock the mutex here before waiting?

Comment (GiedriusS) [reply pkg/store/proxy_heap.go:302]: No, Wait() already takes care of everything for us: https://pkg.go.dev/sync#Cond.Wait.

**Review by fpetkovski — DISMISSED** [review]
Thanks for the quick fix. Would https://github.com/thanos-io/thanos/pull/5742#discussion_r984336548 also get resolved by this?

**Comment by GiedriusS** [comment]
> Thanks for the quick fix. Would [#5742 (comment)](https://github.com/thanos-io/thanos/pull/5742#discussion_r984336548) also get resolved by this?

Yep, I have tested these changes with your PR. It was 100% reproducible before and I couldn't reproduce the nil panic after these changes.

**Review by matej-g — DISMISSED** [review]
The fix is looking good to me, I was just looking at this :+1:

One question related with https://github.com/thanos-io/thanos/pull/5717#issuecomment-1257908216 though, do we then still need to handle 0 response case in https://github.com/thanos-io/thanos/blob/main/pkg/store/proxy_heap.go#L114? I'd want to avoid returning `nil` if that's not a valid case if we are not handling it, to make the code more obvious.

**Comment by GiedriusS** [comment]
> The fix is looking good to me, I was just looking at this +1
> 
> One question related with [#5717 (comment)](https://github.com/thanos-io/thanos/pull/5717#issuecomment-1257908216) though, do we then still need to handle 0 response case in https://github.com/thanos-io/thanos/blob/main/pkg/store/proxy_heap.go#L114? I'd want to avoid returning `nil` if that's not a valid case if we are not handling it, to make the code more obvious.

Good point. Maybe we could even add a `panic()` there? What do you think about that? Because it's pretty much a bug to call At() before Next() and At() shouldn't be called if Next() returns false :smile:

**Comment by matej-g** [comment]
> > The fix is looking good to me, I was just looking at this +1
> > One question related with [#5717 (comment)](https://github.com/thanos-io/thanos/pull/5717#issuecomment-1257908216) though, do we then still need to handle 0 response case in https://github.com/thanos-io/thanos/blob/main/pkg/store/proxy_heap.go#L114? I'd want to avoid returning `nil` if that's not a valid case if we are not handling it, to make the code more obvious.
> 
> Good point. Maybe we could even add a `panic()` there? What do you think about that? Because it's pretty much a bug to call At() before Next() and At() shouldn't be called if Next() returns false smile

Yeah, given that it's really an "invalid case", this solution makes sense to me :+1:

**Comment by matej-g** [comment]
Just ran this also on top #5552, everything is working as expect :ok_hand:, nice job!

**Comment by yeya24** [comment]
This is nice. @GiedriusS We want to tag `v0.28.1` for this?

**Comment by GiedriusS** [comment]
> This is nice. @GiedriusS We want to tag `v0.28.1` for this?

This code is not in 0.28.0 and it's under a hidden flag, by the way (:

**Pull request merged**